xapian-core/cluster/kmeans.cc

```diff
@@ -0,0 +1,139 @@
+/** @file kmeans.cc
+ *  @brief KMeans clustering API
+ */
+/* Copyright (C) 2016 Richhiey Thomas
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+
+#include <config.h>
+
+#include "xapian/cluster.h"
+#include <debuglog.h>
+#include <xapian/error.h>
+#include <omassert.h>
+
+#include <climits>
+#include <random>
+#include <set>
+#include <unordered_map>
+#include <vector>
+
+// Threshold value for checking convergence in KMeans
+#define EPSILON 0.0000000001
+
+using namespace Xapian;
+using namespace std;
+
+KMeans::KMeans(unsigned int k_, unsigned int max_iters_)
+: k(k_), max_iters(max_iters_) {
+    LOGCALL_CTOR(API, "KMeans()", k_ | max_iters_);
+    if (k_ == 0)
+	throw InvalidArgumentError("Number of required clusters should be greater than zero");
+}
+
+string
+KMeans::get_description() const {
+    LOGCALL(API, string, "KMeans::get_description()", NO_ARGS);
+    return "KMeans clusterer";
+}
+
+void
+KMeans::initialize_clusters(ClusterSet &cset) {
+    LOGCALL_VOID(API, "KMeans::initialize_clusters()", cset);
+    int size = docs.size();
+    // Since we are using the Mersenne Twister 19937 pseudo-random number
+    // generator, the period will be greater than the value of k
+    // (period of mt19937 >> k). So we do not need to check whether
+    // the same point has been selected more than once
```

Comment:
@ojwb Sorry for addressing this so late.
Considering what you mentioned about selecting the `i*points.size()/k`, the way we could use random shuffle is by shuffling the numbers between [0 - n-1] and selecting the points the way you mentioned. This would work well because it will give a way to use random as well as rule based selection of initial centroids.
I'm not sure whether this approach preserves the random nature of initial selection, but I'll hold on for comments on this!